src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py

```diff
@@ -0,0 +1,201 @@
+        """
+        Parameters
+        ----------
+        model : torch.nn.module
+            Model to be pruned
+        config_list : list
+            Supported keys:
+                - prune_iterations : The number of rounds for the iterative pruning.
+                - sparsity : The final sparsity when the compression is done.
+        """
+        schema = CompressorSchema([{
+            'sparsity': And(float, lambda n: 0 < n < 1),
+            Optional('op_types'): [str],
+            Optional('op_names'): [str],
+        }], model, _logger)
+
+        schema.validate(config_list)
+
+    def _projection(self, weight, sparsity=0.1):
+        '''
+        Return the Euclidean projection of the weight matrix according to the pruning mode.
+
+        Parameters
+        ----------
+        weight : tensor
+            original matrix
+        sparsity : float
+            the ratio of parameters which need to be set to zero
+
+        Returns
+        -------
+        tensor
+            the projected matrix
+        '''
+        w_abs = weight.abs()
+        if self._pruning_mode == 'fine_grained':
+            k = int(weight.numel() * sparsity)
+            if k == 0:
+                mask_weight = torch.ones(weight.shape).type_as(weight)
+            else:
+                threshold = torch.topk(
+                    w_abs.view(-1), k, largest=False)[0].max()
+                mask_weight = torch.gt(w_abs, threshold).type_as(weight)
+        elif self._pruning_mode == 'channel':
+            filters = weight.size(0)
+            num_prune = int(filters * sparsity)
+            if filters < 2 or num_prune < 1:
+                mask_weight = torch.ones(
+                    weight.size()).type_as(weight).detach()
+            else:
+                w_abs_structured = w_abs.view(filters, -1).sum(dim=1)
+                threshold = torch.topk(w_abs_structured.view(-1),
+                                       num_prune, largest=False)[0].max()
+                mask_weight = torch.gt(w_abs_structured, threshold)[
+                    :, None, None, None].expand_as(weight).type_as(weight)
+
+        return weight.data.mul(mask_weight)
+
+    def get_mask(self, base_mask, weight, num_prune):
+        return L1FilterPruner.get_mask(self, base_mask, weight, num_prune)
+
+    def calc_mask(self, wrapper, **kwargs):
+        """
+        Calculate the mask of given layer.
+        Use the function of LevelPruner or L1FilterPruner according to the pruning mode.
+
+        Parameters
+        ----------
+        wrapper : Module
+            the module to instrument the compression operation
+        Returns
+        -------
+        dict
+            dictionary for storing masks
+        """
+
+        if self._pruning_mode == 'fine_grained':
+            return LevelPruner.calc_mask(self, wrapper, **kwargs)
+        elif self._pruning_mode == 'channel':
+            return L1FilterPruner.calc_mask(self, wrapper, **kwargs)
```

Comment:
`return L1FilterPruner.calc_mask(self, wrapper, **kwargs)` 
This is a nice way to integrate one Pruner into another, but it seems only work for those pruners which don't need intialization,  some pruners such as Slim,  Apoz , TalorOF , need initialization before being used.   Do you have requirement to use such pruners  in another Pruner?

Reply:
The code directory structure is refactored, the `pruning` directory is planned to place the pure pruning algorithms which do not contain any training schedule. We can create another directory to place the Pruners which contains training schedule such as Anneal, Auto compress etc,   we can discuss it later.

Reply:
Refactored with `Masker`.